### PR TITLE
Improved global deadcode elimination

### PR DESCRIPTION
**Pull request by vouillon**
We now take into account of whether an instruction is in a live function or not when generating constraints, so that dead code can no longer result in making a variable live.

We also keep track of nested module.

The performance of the algorithm has been greatly improved as well.

On the `basic_website` example of Tyxml, compiled with `--opt 3`, I get the following reduction in code size:
|     | before | after | |
| ---  | ---: | ---: |  --- |
| uncompressed | 139304 | 125168 | -10% |
| gzip | 47980 | 41367 | -14% |
| bzip2 | 45298 | 38885 | -14% |

Here are the numbers regarding #1612, when compiling `with_bedrock2_fiat_crypto.byte`. For comparison, I have included an intermediate step which just optimizes the current algorithm.
| |before | optimized | after |
| -- | --: | --: | --: |
| duration (s) | 48 | 4.9 | 9.8 |
| peak memory (GiB) | 4 | 2.8 | 2.8 |

### CODE REVIEW

**Comment by hhugo** [comment]
You need to check the effect on https://github.com/ocsigen/js_of_ocaml/issues/1612

**Comment by hhugo** [comment]
> Regarding #1612, when compiling `with_bedrock2_fiat_crypto.byte` the analysis takes 40s with this PR. It takes 48s with `master`. But there is an intermediate step in this PR with an improved solver which reduces the duration of the initial analysis to 15s.

What about memory peak ?

**Comment by vouillon** [comment]
I fixed a memory / speed issue. The numbers are pretty good now.

**Comment by hhugo** [comment]
Cc @jasongross

File: compiler/lib/generate.ml [hunk 1]
```diff
@@ -1124,8 +1124,7 @@ let rec translate_expr ctx queue loc x e level : _ * J.statement_list =
   | Special (Alias_prim name) ->
       let prim = Share.get_prim (runtime_fun ctx) name ctx.Ctx.share in
       (prim, const_p, queue), []
-  | Special Undefined ->
-      (J.(EVar (ident (Utf8_string.of_string_exn "undefined"))), const_p, queue), []
+  | Special Undefined -> (zero, const_p, queue), []
```
Comment (hhugo) [inline compiler/lib/generate.ml:1127]: Maybe rename the Undefined constructor.

**Comment by hhugo** [comment]
I didn't really reviewed https://github.com/ocsigen/js_of_ocaml/pull/1681/commits/df605821096b6f4a357b50c23834c3cc3edc2374 . The rest looks good. Happy to merge.

File: compiler/lib/generate.ml (hunk 1 above)
Comment (vouillon) [reply compiler/lib/generate.ml]: I have now just removed `Special Undefined` and use the integer constant `0` from the start.

**Pull request merged**